include/blmc_robots/real_finger.hpp

```diff
@@ -115,105 +118,187 @@ class RealFinger : public robot_interfaces::Finger {
+                || (summed_velocities.maxCoeff() / SIZE_VELOCITY_WINDOW > STOP_VELOCITY))
+        {
+            Vector torques = -1 * torque_ratio * get_max_torques();
+            TimeIndex t = append_desired_action(torques);
+            Vector abs_velocities = get_observation(t).velocity.cwiseAbs();
+
+            uint32_t running_index = count % SIZE_VELOCITY_WINDOW;
+            if (count >= SIZE_VELOCITY_WINDOW) {
+                summed_velocities -= running_velocities[running_index];
+            }
+            running_velocities[running_index] = abs_velocities;
+            summed_velocities += abs_velocities;
+            count++;
+        }
+
+
+        // At this point we assume that all encoder indices were already
+        // encountered, so the last one seen is the one closest to the end stop
+        // --> no need to move to the index, simply set home position to the
+        // position where the index was last seen.
+        Vector encoder_index_angles = joint_modules_.get_measured_index_angles();
+        if (encoder_index_angles.hasNaN()) {
+            // Make sure we do not try to home on the index when the index
+            // was not found.  This should normally not happen, so no specific
+            // error handling is done here.
+            rt_printf("Homing failed, index not found.  Exit.\n");
+            exit(-1);
+        }
+
+        joint_modules_.set_zero_angles(encoder_index_angles + home_offset);
+
+        rt_printf("Home position: %f, %f, %f\n", encoder_index_angles[0],
+                  encoder_index_angles[1], encoder_index_angles[2]);
     }
 
-    {
 
-      Eigen::Vector3d starting_position;
-      starting_position << 1.5, 1.5, 3.0;
-
-      double kp = 0.4;
-      double kd = 0.0025;
-      int count = 0;
-      Eigen::Vector3d last_diff(std::numeric_limits<double>::max(),
-                                std::numeric_limits<double>::max(),
-                                std::numeric_limits<double>::max());
-      while (count < 2000) {
-        Eigen::Vector3d diff =
-            starting_position - get_observation(current_time_index()).angle;
-
-        // we implement here a small pd control at the current level
-        Eigen::Vector3d desired_torque =
-            kp * diff - kd * get_observation(current_time_index()).velocity;
-
-        // Send the current to the motor
-        TimeIndex t = append_desired_action(desired_torque);
-        wait_until_time_index(t);
-        if (count % 100 == 0) {
-          Eigen::Vector3d diff_difference = last_diff - diff;
-          if (std::abs(diff_difference.norm()) < 1e-5)
-            break;
-          last_diff = diff;
+    /**
+     * @brief Move to given goal position using PD control.
+     *
+     * @param goal_pos Angular goal position for each joint.
+     * @param kp Gain K_p for the PD controller.
+     * @param kd Gain K_d for the PD controller.
+     */
+    void move_to_position(Vector goal_pos, double kp, double kd)
```

Comment:
i think this function should return a boolean indicating whether the position has been reached or not. now it is not clear whether is returns because of timeout, because velocity went to zero, or because it acutally reached the position.

Reply:
done